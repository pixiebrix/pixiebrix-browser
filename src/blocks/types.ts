/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  Expression,
  OutputKey,
  RegistryId,
  TemplateEngine,
  UUID,
} from "@/core";
import { UnknownObject } from "@/types";

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
 * @since 1.4.10
 */
export type URLPattern = string | URLPatternInit;

/**
 * An availability rule. For a rule to match, there must be match from each of the provided entries.
 *
 * An empty value (null, empty array, etc.) matches any site.
 *
 * @see checkAvailable
 */
export type Availability = {
  /**
   * Used to request permissions from the browser
   */
  matchPatterns?: string | string[];
  /**
   * NOTE: the urlPatterns must be a subset of matchPatterns (i.e., more restrictive). If not, PixieBrix may not have
   * access to the page
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
   * @since 1.4.10
   */
  urlPatterns?: URLPattern | URLPattern[];
  /**
   * A selector that must be available on the page in order for the extension to be run.
   *
   * NOTE: the selector must be available at the time the contentScript is installed. While the contentScript is loaded
   * on document_idle, for SPAs this may lead to races between the selector check and rendering of the front-end.
   */
  selectors?: string | string[];
};

/**
 * Availability with consistent shape (i.e., all fields are arrays if provided)
 * @see Availability
 */
export type NormalizedAvailability = {
  matchPatterns?: string[];
  urlPatterns?: URLPattern[];
  selectors?: string[];
};

export type ReaderConfig =
  | RegistryId
  | { [key: string]: ReaderConfig }
  | ReaderConfig[];

/**
 * Where to execute the brick
 * - self: the current tab
 * - opener: the tab that opened the current tab
 * - target: the last tab that the current tab opened
 * - top: the top-most frame in the window
 * - broadcast: all tabs that PixieBrix has access to (the result is returned as an array)
 * - remote: the server (currently only support identity, get, and http bricks)
 * @see {@link BlockConfig.window}
 */
export type BlockWindow =
  | "self"
  | "opener"
  | "target"
  | "top"
  | "broadcast"
  | "remote";

/**
 * Condition expression written in templateEngine for deciding if the step should be run.
 * @see {@link BlockConfig.if}
 */
export type BlockIf = string | boolean | number | Expression;

/**
 * A block configuration to be executed by the PixieBrix runtime.
 * @see runStage
 * @see reducePipeline
 */
export type BlockConfig = {
  /**
   * The registry id of the configured block
   */
  id: RegistryId;

  /**
   * The configuration of the block
   */
  config: UnknownObject;

  /**
   * (Optional) human-readable label for the step.
   *
   * Shown in the page editor, logs, and the progress indicator
   * @see notifyProgress
   */
  label?: string;

  /**
   * (Optional) `true` to indicate on the host page that the step is being run
   */
  notifyProgress?: boolean;

  /**
   * (Optional) If the brick is being run in the context of a deployment, if an error occurs in this step an
   * alert email is send to the admins for the deployment with the brick input.
   *
   * Used as a stopgap/recovery mechanism operations that aren't safe to retry. (For example, sending data to a
   * UiPath queue for execution)
   */
  onError?: {
    alert?: boolean;
  };

  /**
   * Where to execute the brick (default=`self`)
   * @see BlockWindow
   */
  window?: BlockWindow;

  /**
   * The output key (without the preceding "@") to assign the brick output to
   *
   * As of brick API `v2`, the outputKey is required except:
   * - Effect bricks (because they don't produce an output)
   * - The last brick in pipeline
   *
   * @see ApiVersionOptions.explicitDataFlow
   */
  outputKey?: OutputKey;

  /**
   * (Optional) condition expression written in templateEngine for deciding if the step should be run. If not
   * provided, the step is run unconditionally.
   * @see BlockIf
   */
  if?: BlockIf;

  /**
   * (Optional) whether the block should inherit the current root element, or if it should use the document
   * (default=`inherit`)
   *
   * @see root
   * @since 1.4.0
   */
  rootMode?: "inherit" | "document";

  /**
   * (Optional) root jQuery/CSS selector. The selector is relative to the `root` that is passed to the pipeline/stage.
   *
   * An error is thrown at runtime if the selector doesn't match exactly one element
   * @see rootMode
   */
  root?: string;

  /**
   * (Optional) template language to use for rendering the `if` and `config` properties (default=`mustache`)
   * @see config
   * @see if
   */
  templateEngine?: TemplateEngine;

  /**
   * A unique id for the configured block, used to correlate traces across runs when using the Page Editor.
   *
   * DO NOT SET: generated automatically by the Page Editor when configuring a dynamic element.
   */
  instanceId?: UUID;
};

/**
 * A pipeline of blocks to execute sequentially
 */
export type BlockPipeline = BlockConfig[];

/**
 * A control flow branch, for tracing. The array of branches is used to correlate runs of the same block in a pipeline
 * @see TraceMetadata.blockInstanceId
 * @see TraceMetadata.branches
 */
export type Branch = {
  key: string;
  counter: number;
};
