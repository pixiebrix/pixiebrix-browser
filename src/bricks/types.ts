/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type RegistryId } from "@/types/registryTypes";
import {
  type ElementReference,
  type Expression,
  type OutputKey,
  type TemplateEngine,
} from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type UnknownObject } from "@/types/objectTypes";

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
 * @since 1.4.10
 */
type URLPattern = string | URLPatternInit;

/**
 * An availability rule. For a rule to match, there must be match from each of the provided entries.
 *
 * An empty value (null, empty array, etc.) matches any site.
 *
 * @see checkAvailable
 */
export type Availability = {
  /**
   * Used to request permissions from the browser
   */
  matchPatterns?: string | string[];
  /**
   * NOTE: the urlPatterns must be a subset of matchPatterns (i.e., more restrictive). If not, PixieBrix may not have
   * access to the page
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
   * @since 1.4.10
   */
  urlPatterns?: URLPattern | URLPattern[];
  /**
   * A selector that must be available on the page in order for the extension to be run.
   *
   * NOTE: the selector must be available at the time the contentScript is installed. While the contentScript is loaded
   * on document_idle, for SPAs this may lead to races between the selector check and rendering of the front-end.
   */
  selectors?: string | string[];
};

/**
 * Availability with consistent shape (i.e., all fields are arrays if provided)
 * @see Availability
 */
export type NormalizedAvailability = {
  matchPatterns?: string[];
  urlPatterns?: URLPattern[];
  selectors?: string[];
};

export type ReaderConfig =
  | RegistryId
  | { [key: string]: ReaderConfig }
  | ReaderConfig[];

/**
 * Where to execute the brick
 * - self: the current tab/frame
 * - opener: the tab that opened the current tab
 * - target: the last tab that the current tab opened
 * - top: the top-most frame in the window
 * - all_frames: all frames in the window
 * - broadcast: all tabs that PixieBrix has access to (the result is returned as an array)
 * @see {@link BrickConfig.window}
 * @since 1.7.36 - added all_frames
 */
export type BrickWindow =
  | "self"
  | "opener"
  | "target"
  | "top"
  | "broadcast"
  | "all_frames";

/**
 * Returns true if the BrickWindow targets multiple tabs and/or frames.
 *
 * The runtime will return an array of results (one for each successful target)
 */
export function hasMultipleTargets(
  window: BrickWindow
): window is "broadcast" | "all_frames" {
  return window === "broadcast" || window === "all_frames";
}

/**
 * Condition expression written in templateEngine for deciding if the step should be run.
 * @see {@link BrickConfig.if}
 */
export type BlockIf = string | boolean | number | Expression;

/**
 * A brick configuration to be executed by the PixieBrix runtime.
 * @see runStage
 * @see reducePipeline
 */
export type BrickConfig = {
  /**
   * The registry id of the configured brick
   */
  id: RegistryId;

  /**
   * The configuration of the brick
   */
  // XXX: in the future, should use JsonObject here, but will require extra type assertions in the codebase.
  config: UnknownObject;

  /**
   * (Optional) human-readable label for the step.
   *
   * Shown in the page editor, logs, and the progress indicator
   * @see notifyProgress
   */
  label?: string;

  /**
   * (Optional) `true` to indicate on the host page that the step is being run
   */
  notifyProgress?: boolean;

  /**
   * (Optional) If the brick is being run in the context of a deployment, if an error occurs in this step an
   * alert email is sent to the admins for the deployment with the brick input.
   *
   * Used as a stopgap/recovery mechanism operations that aren't safe to retry. (For example, sending data to a
   * UiPath queue for execution)
   */
  onError?: {
    alert?: boolean;
  };

  /**
   * Where to execute the brick (default=`self`)
   * @see BrickWindow
   */
  window?: BrickWindow;

  /**
   * The output key (without the preceding "@") to assign the brick output to
   *
   * As of brick API `v2`, the outputKey is required except:
   * - Effect bricks (because they don't produce an output)
   * - The last brick in pipeline
   *
   * @see ApiVersionOptions.explicitDataFlow
   */
  outputKey?: OutputKey;

  /**
   * (Optional) condition expression written in templateEngine for deciding if the step should be run. If not
   * provided, the step is run unconditionally.
   * @see BlockIf
   */
  if?: BlockIf;

  /**
   * (Optional) whether the brick should inherit the current root element, or if it should use the document
   * (default=`inherit`)
   *
   * `element` rootMode added in 1.7.16
   *
   * @see root
   * @since 1.4.0
   */
  rootMode?: "inherit" | "document" | "element";

  /**
   * (Optional) root jQuery/CSS selector or element reference.
   *
   * For rootMode of "inherit" and "document", interpreted as a jQuery/CSS selector. The selector is relative to the
   * `root` that is passed to the pipeline/stage. An error is thrown at runtime if the selector doesn't match exactly
   * one element
   *
   * @see rootMode
   */
  root?: string | ElementReference | Expression;

  /**
   * (Optional) template language to use for rendering the `if` and `config` properties (default=`mustache`)
   * @see config
   * @see if
   */
  templateEngine?: TemplateEngine;

  /**
   * A unique id for the configured brick, used to correlate traces across runs when using the Page Editor.
   *
   * DO NOT SET: generated automatically by the Page Editor when configuring a dynamic element.
   */
  instanceId?: UUID;
};

/**
 * A pipeline of bricks to execute sequentially
 */
export type BrickPipeline = BrickConfig[];

/**
 * A control flow branch, for tracing. The array of branches is used to correlate runs of the same brick in a pipeline
 * @see TraceMetadata.blockInstanceId
 * @see TraceMetadata.branches
 */
export type Branch = {
  key: string;
  counter: number;
};

/**
 * Defines the position of the brick in the extension
 * ex. "extension.brickPipeline.0.config.body.__value__.0",
 * "extension.brickPipeline.0.config.body.0.children.0.config.onClick.__value__.0"
 */
export type BrickPosition = {
  /**
   * The path to the brick relative to the root pipeline
   */
  path: string;
};
