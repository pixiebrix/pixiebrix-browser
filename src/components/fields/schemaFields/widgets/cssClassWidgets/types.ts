/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Expression } from "@/types/runtimeTypes";
import type React from "react";

export type Spacing = {
  side: string | null;
  size: number;
};

export type Value = string | Expression;

/**
 * An independent class name
 */
export type ClassFlag = {
  /**
   * The Bootstrap 4 class name
   */
  className: string;

  /**
   * Title node to render for the element (in a button/dropdown)
   */
  title: React.ReactNode;

  /**
   * True if the flag is exclusive for it's group (default=true)
   */
  exclusive?: boolean;

  /**
   * Other flags in the same group that the flag implies
   */
  implies?: string[];
};
