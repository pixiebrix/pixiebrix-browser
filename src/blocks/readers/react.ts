/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
