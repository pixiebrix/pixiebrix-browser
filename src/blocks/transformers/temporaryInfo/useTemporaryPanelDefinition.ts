/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useAsyncState } from "@/hooks/common";
import { getPanelDefinition } from "@/contentScript/messenger/api";
import { type UUID } from "@/core";
import { type TemporaryPanelEntry } from "@/sidebar/types";
import { type Target } from "@/types";
import { useEffect, useState } from "react";
import {
  addListener,
  type PanelListener,
  removeListener,
} from "@/blocks/transformers/temporaryInfo/receiverProtocol";
import { validateUUID } from "@/types/helpers";

type PanelDefinition = {
  /**
   * The current panel nonce.
   */
  panelNonce: UUID | null;

  /**
   * The current panel entry.
   */
  entry: TemporaryPanelEntry;
  /**
   * True if the panel definition is being retrieved for the first time.
   */
  isLoading: boolean;
  /**
   * Error if the panel definition could not be retrieved, or null on isLoading or success
   */
  error: unknown;
};

/**
 * Hook to get the panel definition for a given nonce, and watch for definition updates.
 * @param target The target contentScript managing the panel
 * @param initialNonce the initial panelNonce to show
 */
function useTemporaryPanelDefinition(
  target: Target,
  initialNonce: UUID
): PanelDefinition {
  const params = new URLSearchParams(location.search);
  const frameNonce = validateUUID(params.get("frameNonce"));

  const [panelNonce, setNonce] = useState(initialNonce);

  const [entry, isLoading, error, recalculate] = useAsyncState(async () => {
    if (panelNonce) {
      return getPanelDefinition(target, panelNonce);
    }

    // Blank panel, e.g., pre-allocated iframe
    return null;
  }, [panelNonce]);

  useEffect(() => {
    const listener: PanelListener = {
      onUpdateTemporaryPanel(newEntry) {
        console.debug("onUpdateTemporaryPanel", newEntry);
        // Need to verify we're the panel of interest, because messenger broadcasts to all ephemeral panels
        if (newEntry.nonce === panelNonce) {
          // Slightly inefficient to getPanelDefinition because the entry is available in the message. However, this
          // is the cleaner use of useAsyncState
          void recalculate();
        }
      },
      onSetPanelNonce(payload) {
        // Need to verify we're the panel of interest, because messenger broadcasts to all ephemeral panels
        if (
          frameNonce &&
          payload.frameNonce === frameNonce &&
          panelNonce !== payload.panelNonce
        ) {
          // Changing the nonce will cause useAsyncState to recalculate
          console.debug("onSetPanelNonce", payload);
          setNonce(payload.panelNonce);
        }
      },
    };

    addListener(listener);

    return () => {
      removeListener(listener);
    };
  }, [frameNonce, panelNonce, recalculate]);

  return {
    panelNonce,
    entry,
    isLoading,
    error,
  };
}

export default useTemporaryPanelDefinition;
