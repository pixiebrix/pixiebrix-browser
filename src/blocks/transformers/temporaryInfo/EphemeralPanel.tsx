/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import cx from "classnames";
import React, { useEffect } from "react";
import { Button, Modal, Popover } from "react-bootstrap";
import {
  cancelTemporaryPanel,
  resolveTemporaryPanel,
} from "@/contentScript/messenger/api";
import Loader from "@/components/Loader";
import { getErrorMessage } from "@/errors/errorHelpers";
import { type Target } from "@/types/messengerTypes";
import { validateUUID } from "@/types/helpers";
import reportError from "@/telemetry/reportError";
import ErrorBoundary from "@/components/ErrorBoundary";
import PanelBody from "@/sidebar/PanelBody";
import useTemporaryPanelDefinition from "@/blocks/transformers/temporaryInfo/useTemporaryPanelDefinition";
import { type UUID } from "@/types/stringTypes";
import { startCase } from "lodash";
import { type PanelButton } from "@/sidebar/types";
import { ClosePanelAction } from "@/blocks/errors";
import styles from "./EphemeralPanel.module.scss";

type Mode = "modal" | "popover";

const ModalLayout: React.FC<{ className?: string }> = ({
  className,
  children,
}) => (
  // Don't use React Bootstrap's Modal because we want to customize the classes in the layout
  <div className={cx("modal-content", className)}>{children}</div>
);

const PopoverLayout: React.FC<{ className?: string }> = ({
  className,
  children,
}) => (
  // Don't use React Bootstrap's Modal because we want to customize the classes in the layout
  // data-iframe-height is used by iframe-resizer
  <div className={cx("popover", className)} data-iframe-height="">
    {children}
  </div>
);

const ActionToolbar: React.FC<{
  actions: PanelButton[];
  onClick: (action: PanelButton) => void;
}> = ({ actions, onClick }) => (
  <div className={styles.actionToolbar}>
    {actions.map((action, index) => (
      <Button
        key={action.caption ?? action.type ?? index}
        variant={action.variant}
        size="sm"
        onClick={() => {
          onClick(action);
        }}
      >
        {action.caption ?? startCase(action.type)}
      </Button>
    ))}
  </div>
);

/**
 * A modal or popover panel that displays temporary information.
 * @see displayTemporaryInfo
 */
const EphemeralPanel: React.FC = () => {
  const params = new URLSearchParams(location.search);
  const initialNonce: UUID | undefined = validateUUID(params.get("nonce"));
  const opener = JSON.parse(params.get("opener")) as Target;
  const mode = params.get("mode") as Mode;

  // The opener for a sidebar panel will be the sidebar frame, not the host panel frame. The sidebar only opens in the
  // top-level frame, so hard-code the top-level frameId
  const target = opener;

  const Layout = mode === "modal" ? ModalLayout : PopoverLayout;

  const { panelNonce, entry, isLoading, error } = useTemporaryPanelDefinition(
    target,
    initialNonce
  );

  // Report error once
  useEffect(() => {
    if (error) {
      // TODO: https://github.com/pixiebrix/pixiebrix-extension/issues/2769
      reportError(error);
    }
  }, [error]);

  if (isLoading) {
    return (
      <Layout>
        <Loader />
      </Layout>
    );
  }

  if (error) {
    return (
      <Layout>
        <div>Panel Error</div>

        <div className="text-danger my-3">{getErrorMessage(error)}</div>

        <div>
          <Button
            variant="primary"
            onClick={() => {
              cancelTemporaryPanel(target, [panelNonce]);
            }}
          >
            Close
          </Button>
        </div>
      </Layout>
    );
  }

  // Panel was pre-allocated for performance
  if (entry == null) {
    if (mode === "popover") {
      return (
        <Layout>
          <Popover.Title></Popover.Title>
          <Popover.Content>&nbsp;</Popover.Content>
        </Layout>
      );
    }

    return (
      <Layout>
        <Modal.Header></Modal.Header>
        <Modal.Body></Modal.Body>
      </Layout>
    );
  }

  if (mode === "popover") {
    return (
      <Layout>
        <Popover.Title>
          {entry.heading}
          {(entry.showCloseButton ?? true) && (
            <Button
              className="close"
              data-dismiss="alert"
              onClick={() => {
                cancelTemporaryPanel(
                  target,
                  [panelNonce],
                  new ClosePanelAction("User closed the panel")
                );
              }}
            >
              &times;
            </Button>
          )}
        </Popover.Title>
        <Popover.Content>
          <ErrorBoundary>
            <PanelBody
              isRootPanel={false}
              payload={entry.payload}
              context={{
                extensionId: entry.extensionId,
                blueprintId: entry.blueprintId,
              }}
              onAction={(action) => {
                resolveTemporaryPanel(target, panelNonce, action);
              }}
            />
          </ErrorBoundary>

          {entry.actions?.length > 0 && (
            <>
              <hr className={styles.actionDivider} />
              <ActionToolbar
                actions={entry.actions}
                onClick={(action) => {
                  resolveTemporaryPanel(target, panelNonce, action);
                }}
              />
            </>
          )}
        </Popover.Content>
      </Layout>
    );
  }

  return (
    <Layout>
      <Modal.Header
        closeButton={entry.showCloseButton ?? true}
        onHide={() => {
          cancelTemporaryPanel(
            target,
            [panelNonce],
            new ClosePanelAction("User closed the panel")
          );
        }}
      >
        <Modal.Title>{entry.heading}</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <ErrorBoundary>
          <PanelBody
            isRootPanel={false}
            payload={entry.payload}
            context={{
              extensionId: entry.extensionId,
              blueprintId: entry.blueprintId,
            }}
            onAction={(action) => {
              resolveTemporaryPanel(target, panelNonce, action);
            }}
          />
        </ErrorBoundary>
      </Modal.Body>

      {entry.actions?.length > 0 && (
        <Modal.Footer>
          <ActionToolbar
            actions={entry.actions}
            onClick={(action) => {
              resolveTemporaryPanel(target, panelNonce, action);
            }}
          />
        </Modal.Footer>
      )}
    </Layout>
  );
};

export default EphemeralPanel;
