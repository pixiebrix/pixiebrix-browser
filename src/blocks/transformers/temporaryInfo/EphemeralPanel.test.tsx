/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@testing-library/react";
import EphemeralPanel from "@/blocks/transformers/temporaryInfo/EphemeralPanel";
import { uuidv4 } from "@/types/helpers";
import useTemporaryPanelDefinition from "@/blocks/transformers/temporaryInfo/useTemporaryPanelDefinition";
import { waitForEffect } from "@/testUtils/testHelpers";
import {
  cancelTemporaryPanel,
  resolveTemporaryPanel,
} from "@/contentScript/messenger/api";
import { sidebarEntryFactory } from "@/testUtils/factories";

jest.mock(
  "@/blocks/transformers/temporaryInfo/useTemporaryPanelDefinition",
  () => ({
    __esModule: true,
    default: jest.fn(),
  })
);

jest.mock("@/contentScript/messenger/api", () => ({
  cancelTemporaryPanel: jest.fn(),
  resolveTemporaryPanel: jest.fn(),
}));

jest.mock("@/sidebar/PanelBody", () => ({
  __esModule: true,
  default: jest.fn(() => <div data-testid="panel-body"></div>),
}));

const useTemporaryPanelDefinitionMock =
  useTemporaryPanelDefinition as jest.MockedFunction<
    typeof useTemporaryPanelDefinition
  >;

const cancelTemporaryPanelMock = cancelTemporaryPanel as jest.MockedFunction<
  typeof cancelTemporaryPanel
>;
const resolveTemporaryPanelMock = resolveTemporaryPanel as jest.MockedFunction<
  typeof resolveTemporaryPanel
>;

describe("EphemeralPanel", () => {
  beforeEach(() => {
    useTemporaryPanelDefinitionMock.mockReset();
    cancelTemporaryPanelMock.mockReset();
    resolveTemporaryPanelMock.mockReset();
  });

  test.each(["modal", "popover"])("renders blank: %s", async (mode) => {
    location.href = `chrome-extension://abc/ephemeralPanel.html?mode=${mode}&frameNonce=${uuidv4()}`;

    useTemporaryPanelDefinitionMock.mockReturnValue({
      panelNonce: null,
      entry: null,
      error: null,
      isLoading: null,
    });

    const rendered = render(<EphemeralPanel />);

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test.each(["modal", "popover"])("renders error state: %s", async (mode) => {
    location.href = `chrome-extension://abc/ephemeralPanel.html?mode=${mode}&frameNonce=${uuidv4()}`;

    useTemporaryPanelDefinitionMock.mockReturnValue({
      panelNonce: uuidv4(),
      entry: null,
      error: new Error("test error"),
      isLoading: null,
    });

    const rendered = render(<EphemeralPanel />);

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();

    rendered.queryByText("Close").click();

    expect(cancelTemporaryPanelMock).toHaveBeenCalled();
  });

  test.each(["modal", "popover"])(
    "renders action buttons: %s",
    async (mode) => {
      location.href = `chrome-extension://abc/ephemeralPanel.html?mode=${mode}&frameNonce=${uuidv4()}`;

      const panelNonce = uuidv4();

      useTemporaryPanelDefinitionMock.mockReturnValue({
        panelNonce,
        entry: sidebarEntryFactory("temporaryPanel", {
          nonce: panelNonce,
          actions: [{ type: "testClick", variant: "light" }],
        }),
        error: null,
        isLoading: null,
      });

      const rendered = render(<EphemeralPanel />);

      await waitForEffect();

      expect(rendered.asFragment()).toMatchSnapshot();

      rendered.queryByText("Test Click").click();

      expect(cancelTemporaryPanelMock).not.toHaveBeenCalled();
      expect(resolveTemporaryPanelMock).toHaveBeenCalledWith(null, panelNonce, {
        type: "testClick",
        variant: "light",
      });
    }
  );
});
