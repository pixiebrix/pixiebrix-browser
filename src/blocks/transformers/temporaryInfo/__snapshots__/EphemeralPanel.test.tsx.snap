// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EphemeralPanel renders action buttons: modal 1`] = `
<DocumentFragment>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Test Title
      </div>
      <button
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          class="sr-only"
        >
          Close
        </span>
      </button>
    </div>
    <div
      class="modal-body"
    >
      <div
        data-test-id="panel-body"
      />
    </div>
    <div
      class="modal-footer"
    >
      <div
        class="actionToolbar"
      >
        <button
          class="btn btn-light btn-sm"
          type="button"
        >
          Test Click
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EphemeralPanel renders action buttons: popover 1`] = `
<DocumentFragment>
  <div
    class="popover"
    data-iframe-height=""
  >
    <div
      class="popover-header"
    >
      Test Title
      <button
        class="close btn btn-primary"
        data-dismiss="alert"
        type="button"
      >
        ×
      </button>
    </div>
    <div
      class="popover-body"
    >
      <div
        data-test-id="panel-body"
      />
      <hr
        class="actionDivider"
      />
      <div
        class="actionToolbar"
      >
        <button
          class="btn btn-light btn-sm"
          type="button"
        >
          Test Click
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EphemeralPanel renders blank: modal 1`] = `
<DocumentFragment>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    />
    <div
      class="modal-body"
    />
  </div>
</DocumentFragment>
`;

exports[`EphemeralPanel renders blank: popover 1`] = `
<DocumentFragment>
  <div
    class="popover"
    data-iframe-height=""
  >
    <div
      class="popover-header"
    />
    <div
      class="popover-body"
    >
       
    </div>
  </div>
</DocumentFragment>
`;

exports[`EphemeralPanel renders error state: modal 1`] = `
<DocumentFragment>
  <div
    class="modal-content"
  >
    <div>
      Panel Error
    </div>
    <div
      class="text-danger my-3"
    >
      test error
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EphemeralPanel renders error state: popover 1`] = `
<DocumentFragment>
  <div
    class="popover"
    data-iframe-height=""
  >
    <div>
      Panel Error
    </div>
    <div
      class="text-danger my-3"
    >
      test error
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</DocumentFragment>
`;
